### PR TITLE
Eip 1344 (ChainID opcode)

### PR DESCRIPTION
**Pull request by holiman**
Implementation of 1344. This PR is built on top of https://github.com/ethereum/go-ethereum/pull/19743 , which contains some necessary framework. 
Neither of these PRs actually enable it for Istanbul (that needs a follow-up PR). So in that sense, they should both be fairly safe to merge. 
cc @fubuloubu  (only the last commit is relevant)

### CODE REVIEW

File: core/vm/eips.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		constantGas: GasQuickStep,
+		minStack:    minStack(0, 1),
+		maxStack:    maxStack(0, 1),
+		valid:       true,
+	}
+}
+
+// opChainID implements CHAINID opcode
+func opChainID(pc *uint64, interpreter *EVMInterpreter, contract *Contract, memory *Memory, stack *Stack) ([]byte, error) {
+	chainId := interpreter.intPool.get().Set(interpreter.evm.chainConfig.ChainID)
```
Comment (fubuloubu) [inline core/vm/eips.go]: `interpreter.evm.chainConfig.ChainID` is set via chain configuration variable and given to the VM? If so, this is correct behavior.

Comment (holiman) [reply core/vm/eips.go]: Yes, it is

File: core/vm/interpreter.go [hunk 2]
```diff
@@ -88,20 +91,29 @@ func NewEVMInterpreter(evm *EVM, cfg Config) *EVMInterpreter {
...
+			jt = homesteadInstructionSet
 		default:
-			cfg.JumpTable = frontierInstructionSet
+			jt = frontierInstructionSet
+		}
+		for i, eip := range cfg.ExtraEips {
+			if err := EnableEIP(eip, &jt); err != nil {
+				// Disable it, so caller can check if it's activated or not
+				cfg.ExtraEips = append(cfg.ExtraEips[:i], cfg.ExtraEips[i+1:]...)
+				log.Error("eip %d failed activation: %v", err)
```
Comment (CoreyLin) [inline core/vm/interpreter.go]: Should it be modified to the following code?

`log.Error(fmt.Sprintf("eip %d failed activation: %v", eip, err))`

Comment (holiman) [reply core/vm/interpreter.go]: Absolutely, thanks!

Comment (holiman) [reply core/vm/interpreter.go]: I fixed it on the upstream PR, and now rebased on top of the fixed one

File: core/vm/opcodes.go [hunk 3]
```diff
@@ -101,6 +101,8 @@ const (
 	NUMBER
 	DIFFICULTY
 	GASLIMIT
+	CHAINID     = 0x46
```
Comment (CoreyLin) [inline core/vm/opcodes.go]: Maybe "= 0x46" and "= 0x47" could be omitted because of iota effect.

File: tests/state_test_util.go [hunk 4]
```diff
@@ -122,10 +146,11 @@ func (t *StateTest) Subtests() []StateSubtest {
 
 // Run executes a specific subtest.
 func (t *StateTest) Run(subtest StateSubtest, vmconfig vm.Config) (*state.StateDB, error) {
-	config, ok := Forks[subtest.Fork]
-	if !ok {
+	config, eips, err := getVMConfig(subtest.Fork)
+	if err != nil {
```
Comment (CoreyLin) [inline tests/state_test_util.go]: Maybe it could be modified to:

    if config, eips, err := getVMConfig(subtest.Fork); err != nil {

**Comment by karalabe** [comment]
Pls rebase.

File: core/vm/opcodes.go (hunk 3 above)
Comment (holiman) [reply core/vm/opcodes.go]: They could, but I dislike the iota effect -- imo it's wrong to use iota to define something that is defined _externally_, and it should only be used for things where there's no external dependency -- where only internal consistency matters

**Comment by holiman** [comment]
rebased

File: core/vm/eips.go [hunk 5]
```diff
@@ -61,3 +63,23 @@ func opSelfBalance(pc *uint64, interpreter *EVMInterpreter, contract *Contract,
 	stack.push(balance)
 	return nil, nil
 }
+
+// Enable1344 applies EIP-1344 (ChainID Opcode)
+// - Aadds an opcode that returns the current chain’s EIP-155 unique identifier
```
Comment (rjl493456442) [inline core/vm/eips.go]: Does `Aadds` means `and adds`?

Comment (karalabe) [reply core/vm/eips.go]: Also lowercase e on the method doc.

Comment (holiman) [reply core/vm/eips.go]: No, it means `Adds`

File: core/vm/eips.go [hunk 6]
```diff
@@ -61,3 +63,23 @@ func opSelfBalance(pc *uint64, interpreter *EVMInterpreter, contract *Contract,
 	stack.push(balance)
 	return nil, nil
 }
+
+// Enable1344 applies EIP-1344 (ChainID Opcode)
```
Comment (holiman) [inline core/vm/eips.go]:
```suggestion
// enable1344 applies EIP-1344 (ChainID Opcode)
```

File: core/vm/eips.go (hunk 5 above)
Comment (holiman) [inline core/vm/eips.go]:
```suggestion
// - Adds an opcode that returns the current chain’s EIP-155 unique identifier
```

**Pull request merged**